src/magic_folder/magic_folder.py

```diff
@@ -175,6 +179,36 @@ def ready(self):
         """
         return defer.succeed(None)
 
+    def participants(self):
+        # type: () -> Deferred[list[IParticipant]]
+        """
+        List all participants of this folder
+        """
+        return self._participants.list()
+
+    def add_participant(self, author, participant_directory):
+        return self._participants.add(author, participant_directory)
+
+    def add_snapshot(self, relative_path):
+        # type: (unicode) -> Deferred[None]
+        """
+        Create a new snapshot of the given file.
+        """
+
+        # preauthChild allows path-separators in the "path" (i.e. not
+        # just a single path-segment). That is precisely what we want
+        # here, though. It sill does not allow the path to "jump out"
+        # of the base magic_path -- that is, an InsecurePath error
+        # will result if you pass an absolute path outside the folder
+        # or a relative path that reaches up too far.
+        try:
+            path = self.config.magic_path.preauthChild(relative_path)
+        except InsecurePath as e:
+            return defer.fail(
+                APIError.from_exception(http.NOT_ACCEPTABLE, e)
```

Comment:
The ticket says you want to remove tahoe-code from the Web code .. but now there's Web code mixed into the magic-folder code; this should raise something other than a web-specific exception.

Reply:
Yes, it's the HTTP error-code that seems web-specific. I guess if we're _already_ scattering HTTP codes throughout other code then this is fine.

Reply:
Yeah, we've already done that. I'd prefer to not have that, but I think (at least until the code base is more settled), having that at the location of the error makes more sense. The other option would be to have a bunch of hyper-specific exception classes, and map them to error codes either in the exception (which is *really* all that different than what we have now) or have some sort of table in the web code. The latter feels like unnecessary maintenance overhead at this point.

We can do things like https://github.com/LeastAuthority/magic-folder/blob/039a99443b4f0404bf27d0c786e62448579e1203/src/magic_folder/common.py#L109-L127 if there are common errors, but I think most of the explicit uses of `APIError` are all more-or-less one-offs.